lib/thread.c

```diff
@@ -41,6 +41,14 @@ DEFINE_MTYPE_STATIC(LIB, THREAD_STATS,  "Thread stats")
 #include <mach/mach_time.h>
 #endif
 
+static unsigned char wakebyte = 0x01;
+
+#define ACQUIRE_MASTER(m) \
+  do { \
+    while (pthread_mutex_trylock (&m->mtx)) \
+      write (m->io_pipe[1], &wakebyte, 1); \
+  } while (0);
+
```

Comment:
>If you make it nonblocking it will call write() for 1829410929 times until the kernel decides to schedule the other thread and actually make progress on the select()...It would in fact be better to change the code to write 4096 bytes and keep it blocking. That way there's 2 write calls at worst, the second of which will immediately block. It'll still be less overhead than writing 4096 * 1 byte.

Ah yep, this is much better.

>Copy it under lock, then release the lock. The select() call does already copy it (because fd_set is a value type), it's just the poll() path that has the issue.

I was thinking there would be issues if another pthread came in and cancelled an I/O task during poll. Then that pthread would expect that the associated task wouldn't execute after that point, but poll() could still return the fd. After that point, since the associated thread has been removed from the fd-indexed array, the program either runs the wrong thread or segfaults. But since we've decided thread_cancel() is mt-unsafe that's acceptable. I'll make the changes.